// @algorandfoundation/algokit-utils/types/amount.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlgoAmount = void 0;
const algosdk_1 = __importDefault(require("algosdk"));
/** Wrapper class to ensure safe, explicit conversion between µAlgos, Algos and numbers */
class AlgoAmount {
    /** Return the amount as a number in µAlgos */
    get microAlgos() {
        return this.amountInMicroAlgos;
    }
    /** Return the amount as a number in Algos */
    get algos() {
        return algosdk_1.default.microalgosToAlgos(this.amountInMicroAlgos);
    }
    constructor(amount) {
        this.amountInMicroAlgos = 'microAlgos' in amount ? amount.microAlgos : algosdk_1.default.algosToMicroalgos(amount.algos);
    }
    toString() {
        return `${this.microAlgos.toLocaleString('en-US')} µALGO${this.microAlgos === 1 ? '' : 's'}`;
    }
    valueOf() {
        return this.microAlgos;
    }
    /** Create a @see {AlgoAmount} object representing the given number of Algos */
    static Algos(amount) {
        return new AlgoAmount({ algos: amount });
    }
    /** Create a @see {AlgoAmount} object representing the given number of µAlgos */
    static MicroAlgos(amount) {
        return new AlgoAmount({ microAlgos: amount });
    }
}
exports.AlgoAmount = AlgoAmount;
