// @algorandfoundation/algokit-utils/types/indexer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountStatus = exports.SignatureType = exports.ApplicationOnComplete = void 0;
var ApplicationOnComplete;
(function (ApplicationOnComplete) {
    ApplicationOnComplete["noop"] = "noop";
    ApplicationOnComplete["optin"] = "optin";
    ApplicationOnComplete["closeout"] = "closeout";
    ApplicationOnComplete["clear"] = "clear";
    ApplicationOnComplete["update"] = "update";
    ApplicationOnComplete["delete"] = "delete";
})(ApplicationOnComplete = exports.ApplicationOnComplete || (exports.ApplicationOnComplete = {}));
var SignatureType;
(function (SignatureType) {
    SignatureType["sig"] = "sig";
    SignatureType["msig"] = "msig";
    SignatureType["lsig"] = "lsig";
})(SignatureType = exports.SignatureType || (exports.SignatureType = {}));
var AccountStatus;
(function (AccountStatus) {
    AccountStatus["Offline"] = "Offline";
    AccountStatus["Online"] = "Online";
    AccountStatus["NotParticipating"] = "NotParticipating";
})(AccountStatus = exports.AccountStatus || (exports.AccountStatus = {}));
